### ISSUE TITLE
Feedforward neural network model optimization

### ISSUE THREAD

**Issue by afonsocastro** [afonsocastro/larcc_interface#10]
### **Optimization 1**

**Searched combinations:**
n_hidden_layers = 2 or 3
neurons_in_layer = 32 or 64
dropout = 0 or 0.2
activation_function = _relu_ or _sigmoid_ or _selu_
model_optimizer = _Adam_ or _SGD_ or _RMSprop_

**Fixed values:**
learning_rate = 0.001
max_epochs = 300
batch_size = 96
model_loss = _sparse_categorical_crossentropy_ 
kernel_regularizer_per_layer = _L1_
early_stopping = _val_loss_ ; patience = 20

**model configurations:** 5616
**searched time:** 9h16m36s

### BEST Result
![image](https://user-images.githubusercontent.com/47828797/181111684-5267ca47-4215-43d2-b8c6-6b8f5c97e95d.png)
![image](https://user-images.githubusercontent.com/47828797/181111702-15903be3-0da8-436f-8e03-022e8d61995d.png)
![image](https://user-images.githubusercontent.com/47828797/181111719-4b1a976b-2ae6-464a-8fc4-ab20a3cab103.png)

**Epoch 300/300:**
loss: 0.4321 - accuracy: 0.9745
val_loss: 0.4490 - val_accuracy: 0.9583

Using _784 samples for training_ and _336 for validation_

**10 best results (optimization)**
![image](https://user-images.githubusercontent.com/47828797/181111902-6ee34a2b-2452-4733-8c98-1e5f36adb3c4.png)

**Comment by vitoruapt** [comment]
The curves seem convincing! Can we conclude that we have reached a good architecture? The val_accuracy is nice (~96%), although I would expect a little better loss (a bit smaller wouldn't harm).
What troubles me is that among the 10 best optimization trials, we have similar accuracies for relatively different architectures, both with 2 and 3 hidden layers, and with quite disparate number of epochs (184 -- 300). Perhaps the problem is not that hard, and there are many ways to solve it. I say, let's expand the dataset to other users and test this network with it!

**Comment by afonsocastro** [comment]
### **Optimization 2**

**Searched combinations:**

neurons_in_layer = 16 or 32 or 64 or 128
dropout = 0 or 0.2
activation_function = _relu_ or _selu_

**Fixed values:**
n_hidden_layers = 3
model_optimizer = _Adam_
learning_rate = 0.001
max_epochs = 300
batch_size = 96
model_loss = _sparse_categorical_crossentropy_ 
kernel_regularizer_per_layer = _L1_
early_stopping = _val_loss_ ; patience = 20

**model configurations:** 4096
**searched time:** 6h52m26s

### BEST Result
![image](https://user-images.githubusercontent.com/47828797/181299442-a55c4d80-20e3-4726-bb32-b8e1f44336bb.png)
![image](https://user-images.githubusercontent.com/47828797/181299483-d0d9b632-9d33-4e1c-a58e-e2a3bd013a93.png)
![image](https://user-images.githubusercontent.com/47828797/181299502-54fb65c5-16d4-48f5-955b-0de855005e69.png)


**Epoch 300/300:**
loss: 0.4050 - accuracy: 0.9834
val_loss: 0.4435 - val_accuracy: 0.9554

Using _784 samples for training_ and _336 for validation_

**10 best results (optimization)**
![image](https://user-images.githubusercontent.com/47828797/181299644-e3985c31-3a1f-4ed8-99f9-e4bcbc9d639b.png)

**Comment by afonsocastro** [comment]
> The curves seem convincing! Can we conclude that we have reached a good architecture? The val_accuracy is nice (~96%), although I would expect a little better loss (a bit smaller wouldn't harm).
> What troubles me is that among the 10 best optimization trials, we have similar accuracies for relatively different architectures, both with 2 and 3 hidden layers, and with quite disparate number of epochs (184 -- 300). Perhaps the problem is not that hard, and there are many ways to solve it. I say, let's expand the dataset to other users and test this network with it!

From the first optimization, I can say that the only reliable architecture with only 2 hidden layers, is the one with the dropout of 20% at the second layer. The other 2-layer architecture EarlyStopped at the 184 epoch, which I do not find trustworthy.

Fact: On the 2 groups of _10 best results_ of both optimizations, we never saw any configuration with dropout in more than just one layer. Either there is a dropout on only one layer, or there is no dropout at all.

> Perhaps the problem is not that hard, and there are many ways to solve it.

I completely agree with it, since we outcome several acceptable model configurations with any value of neurons (16, 32, 64 or 128) and for both activation functions (relu and selu), with or without dropout. 

However, we are going to create a new dataset, in which the joints efforts are not included any more. This means that instead of having several groups of 13 values (timestamp + 6 fist force/torque + 6 joints torques), we will now have several groups of only 7 values (timestamp + 6 fist force/torque). 
This probably will bring some differences to the new optimal neural network architecture.

**Comment by afonsocastro** [comment]
### **Optimization 3**

**Searched combinations:**
n_hidden_layers = 3 or 4
neurons_in_layer = 64 or 128
dropout = 0 or 0.5
activation_function = _relu_ or _selu_

**Fixed values:**
model_optimizer = _Adam_
learning_rate = 0.001
max_epochs = 300
batch_size = 96
model_loss = _sparse_categorical_crossentropy_ 
kernel_regularizer_per_layer = _L1_
early_stopping = _val_loss_ ; patience = 20

**model configurations:** 4608
**searched time:** 7h14m17s

### BEST Result
![image](https://user-images.githubusercontent.com/47828797/181618861-f010946a-4385-4df0-b74f-5e071cc5736c.png)
![image](https://user-images.githubusercontent.com/47828797/181618880-2f9c903b-fb0d-460b-a891-0593cdcc5e89.png)
![image](https://user-images.githubusercontent.com/47828797/181618894-44910054-d8f0-465c-9aaf-9feb03e7d657.png)



**Epoch 249/300:**
loss: 0.5427 - accuracy: 0.9566
val_loss: 0.5808 - val_accuracy: 0.9405

Using _784 samples for training_ and _336 for validation_

**10 best results (optimization)**
![image](https://user-images.githubusercontent.com/47828797/181619010-07f610a6-cf73-4870-a221-2a8b3be8aeba.png)

**Comment by afonsocastro** [comment]
The last Optimization (Optimization 3) was the only one where the best configuration could never achieve the 300 epoch.

With all these different studies, I will propose the following conjectures:

1. the model architecture should have 3 hidden layers
2. the dropout option should be only available on the 3rd (last) hidden layer
3. the neurons activation function should be _relu_ or _selu_

I will try a new optimization with this information in mind.

**Comment by afonsocastro** [comment]
### **Optimization 4**

**Searched combinations:**
neurons_in_layer = 16 or 32 or 64 or 128
dropout = 0 or 0.2 or 0.5
activation_function = _relu_ or _selu_

**Fixed values:**
n_hidden_layers = 3
model_optimizer = _Adam_
learning_rate = 0.001
max_epochs = 300
batch_size = 96
model_loss = _sparse_categorical_crossentropy_ 
kernel_regularizer_per_layer = _L1_
early_stopping = _val_loss_ ; patience = 20

**model configurations:** 1536
**searched time:** 2h34m41s

### BEST Result
![image](https://user-images.githubusercontent.com/47828797/181767883-c756dcb7-8630-47d6-bfa4-3e8f717ed44e.png)
![image](https://user-images.githubusercontent.com/47828797/181767904-e4314b49-9313-4526-9b9b-63b4e0342346.png)
![image](https://user-images.githubusercontent.com/47828797/181767922-fdef7b06-a58b-4bc1-8613-f885d58e36dd.png)


**Epoch 300/300:**
loss: 0.4666 - accuracy: 0.9745
val_loss: 0.5030 - val_accuracy: 0.9524

Using _784 samples for training_ and _336 for validation_

**10 best results (optimization)**
![image](https://user-images.githubusercontent.com/47828797/181767944-ebcb229f-2a68-4ba5-aeac-8d4ad716558f.png)

**Comment by afonsocastro** [comment]
I will now run the following optimization:

**Searched combinations:**
n_hidden_layers = 1 or 2 or 3
neurons_in_layer = 16 or 32 or 64
dropout (on last layer) = 0 or 0.2 or 0.5
learning_rate = 0.01 or 0.001 or 0.0001
activation_function = _relu_ or _selu_ or _softsign_ or _tanh_

**Fixed values:**
model_optimizer = Adam
max_epochs = 500
batch_size = 96
model_loss = sparse_categorical_crossentropy
kernel_regularizer_per_layer = L1
early_stopping = val_loss ; patience = 20


**Note**: I will run each different architecture 3 times, and extract the mean of the 3  values of val_accuracy and last_epoch:
model configurations: 16956
number of trained networks (tests): 50868 = 16956 x 3

**Comment by afonsocastro** [comment]
With the new dataset collected on 1 August 2022 (see [this post](https://human-robot-collaboration.blogspot.com/2022/08/data-acquisition-experiment.html)), I just optimized some parameters to achieve a proper and valid neural network.

### Optimization 5

**Searched combinations:**
n_hidden_layers = 2 or 3
neurons_in_layer = 32 or 64
dropout = 0.2 or 0.5
activation_function = relu or selu

**Fixed values:**
model_optimizer = Adam
learning_rate = 0.001
max_epochs = 300
batch_size = 96
model_loss = sparse_categorical_crossentropy
kernel_regularizer_per_layer = L1
early_stopping = val_loss ; patience = 30

model configurations: 160
searched time: 16m55s

### BEST Result
![image](https://user-images.githubusercontent.com/47828797/188671782-06300b49-905a-4b6c-919d-996886efdf3e.png)

![image](https://user-images.githubusercontent.com/47828797/188680286-407ecab8-069f-4f5d-98e0-66afe8b3cc49.png)
![image](https://user-images.githubusercontent.com/47828797/188680392-d883fb9f-6e0a-411e-9513-46c92be02084.png)

**Epoch 218/300:**
loss: 0.7502 - accuracy: 0.8859
val_loss: 0.7890 - val_accuracy: 0.8566

Using 1253 samples for training and 537 for validation

### Confusion Matrix
![image](https://user-images.githubusercontent.com/47828797/188680960-c3774169-af63-4f52-ae5a-6c956800ffe8.png)

Using 768 new samples for predicting

**10 best results (optimization)**
![image](https://user-images.githubusercontent.com/47828797/188671655-7d336128-804e-4d6e-bca6-928809f92e94.png)
